### PR TITLE
Proposed changes to enforce proper ordering of pseudo-headers.

### PR DESCRIPTION
**Pull request by byronformwalt**
Ilya, here are my recommended changes.  I also updated to latest Rubocop gem and eliminated new violations either by disabling them or making appropriate style/layout changes.  All tests are passing on my machine.

### CODE REVIEW

**Comment by ioquatix** [comment]
This will beak existing code where `headers` is not an array or hash or otherwise doesn't respond to `#sort`.

**Comment by ioquatix** [comment]
There are two solutions IMHO.

1. Call sort on the headers before supplying it to the stream
2. Use an array or other data structure and ensure that headers show in the required order.

**Comment by ioquatix** [comment]
Here is what `async-http` does:

https://github.com/socketry/async-http/blob/f3f984999835ae8b195cd6b7aaafae2ed8968a12/lib/async/http/protocol/http2.rb#L334-L339

Here is the implementation of `Merged`: https://github.com/socketry/async-http/blob/f3f984999835ae8b195cd6b7aaafae2ed8968a12/lib/async/http/headers.rb#L188-L200

This guarantees the following:

1. Order of `:` headers comes first.
2. No merge/sort or other operation on supplied hash is more efficient.
3. User can supply headers in any data structure they want, it just needs to respond to `#each do |key, value|`.

**Comment by ioquatix** [comment]
No I wouldn't really be enthusiastic about such a solution. It's unnecessary inefficient both in execution and memory allocations.

I don't believe the solution to this problem belongs in this library, I believe that it's the responsibility of the caller to ensure the headers are in the correct order.

The correct solution is to change `net-http2` to do the right thing (i.e. calling `sort` if that is appropriate) or use a similar technique to `async-http` `Merged`.

In a more abstract sense, I think that it's not up to `http-2` to enforce correct behaviour. It's job is to provide an API for interacting with HTTP/2 connections and streams. There are lots of things you can do incorrectly.

As an example of where you might want this behaviour - let's say you want to use this library to test whether upstream implementations are correct - then you will need the ability to specify headers in the wrong order to check errors.

This library is a wrapper around the protocol, not a high level request/response library. It's expected such a high level library will need to use the protocol in the correct way.

**Comment by byronformwalt** [comment]
I share your sentiment on efficiency.  However, I think that if http-2 encounters a GOAWAY frame with debug info returned by the server, it should at least raise an exception with that bearing that content.

**Comment by ioquatix** [comment]
There is a fine line between things which are exceptional, and things which are normal behaviour of the protocol. I don't think `GOAWAY` fits the former.

**Comment by HoneyryderChuck** [comment]
Agreed, user code should sort then pass headers. Also, making so many changes in the same PR which are out of scope is not the ideal approach.

**Comment by ioquatix** [comment]
@byronformwalt thanks for your effort. Do you think you can reference further issues you create in `net-http2` so we can track them here too?

File: spec/compressor_spec.rb [hunk 1]
```diff
@@ -494,12 +495,13 @@
             end
             it 'should emit expected headers' do
               subject
-              # order-perserving compare
-              expect(@emitted).to eq ex[:streams][nth][:emitted]
+              # Sorted compare (pseudo-headers should always be first.)
```
Comment (igrigorik) [inline spec/compressor_spec.rb]: We should add a standalone spect that tests the `sort` invariant.

**Comment by igrigorik** [comment]
@ioquatix but we *do* expect `headers` to [be an Array](https://github.com/igrigorik/http-2/blob/3457e25f0ce4addd6cf80d9335c1a618e0a2de90/lib/http/2/compressor.rb#L196), and we count on the fact that it responds to `each`. I don't think it's an unreasonable leap to also assume that it must respond to `sort`. 

Given that the behavior we're discussing here is a normative `MUST` [in the h2 protocol](url), I do think that this is something we should handle and enforce in the library:

> All pseudo-header fields MUST appear in the header block before regular header fields. Any request or response that contains a pseudo-header field that appears in a header block after a regular header field MUST be treated as malformed (Section 8.1.2.6).

WDYT?

**Comment by ioquatix** [comment]
@igrigorik Thanks for following up with a specific reference to the spec, I appreciate that.

> @ioquatix but we do expect headers to be an Array, and we count on the fact that it responds to each. I don't think it's an unreasonable leap to also assume that it must respond to sort.

The only expectation up until this point is that the passed in object responds to `#each {|key,value|`. `async-http` now depends on that. I could change it, but I actually like that design since it gives a lot of flexibility. Other libraries may or may not depend on this already.

> Given that the behavior we're discussing here is a normative MUST in the h2 protocol, I do think that this is something we should handle and enforce in the library.

- Sort is not free. For large number of headers it could be quite expensive, and unexpected cost.
- If we don't change the API, we need to internally convert to an array, and then sort that. It's potentially a number of memory allocations.
- It's feels like a side effect of sort that `:` comes first. `Array#sort` will actually sort based on the entire `[key, value]`, not just the key as you might expect.
- It changes the order of headers even when it is already in the necessary order.
- It's trivial for client library to do the right thing according to their own needs.

Additionally, I thought about throwing an exception if the headers are not in order.

However, this prevents the library from being used to test this particular part of the spec. For example, it will become impossible to send a request to a remote server to determine if it will return 400 bad request. This is something I do a lot in `async-http`, i.e. test both positive and negative cases. It is likely I will add this invariant into the server side of `async-http` and return 400 bad request. I will need to be able to supply headers in incorrect order to test this.

File: lib/http/2/compressor.rb [hunk 2]
```diff
@@ -550,7 +550,7 @@ def header(buf)
       def decode(buf)
         list = []
         list << @cc.process(header(buf)) until buf.empty?
-        list.compact
+        list.compact.sort
```
[the remark this answers is not in the document]
Comment (byronformwalt) [reply lib/http/2/compressor.rb]: @igrigorik Yes, sorry about that.  The only reason I injected that modification was to enable the tests to pass.  However, when I later modified the test suite to perform a sorted comparison, I failed to retract the sort operation I added into the decoder.

**Comment by byronformwalt** [comment]
@igrigorik Given that a traditional sort would not be performant, and that the client is required to transmit pseudo headers first, I propose the following performant partitioning solution as a compromise:

```
module HTTP2
  module Header
    class Compressor
      def encode(mixed_headers)
        buffer = Buffer.new
        pseudo_headers,headers = mixed_headers.partition{|field,value| field[0] == ':'}
        commands = @cc.encode(pseudo_headers + headers)  
        commands.each do |cmd|
          buffer << header(cmd)
        end

        buffer
      end
    end
  end
end
```

What do you think?

**Comment by igrigorik** [comment]
@byronformwalt @ioquatix why the big concern with `sort`? We're handling collections on the order of tens of entries; I'm not convinced that the performance overhead of quicksort will be noticeable. Is there a particular scenario or benchmark that we're afraid of? 

> Additionally, I thought about throwing an exception if the headers are not in order.

Personally, I think we should handle this in http-2. This is a non-obvious requirement deep in the guts of the protocol and we should do the right thing (tm) on user's behalf.

**Comment by ioquatix** [comment]
I would be okay with @HoneyryderChuck proposal, but that limits my ability to specifically test this case, which is a pain because I do want to add specs for it, so I can check my sever returns 400 bad request.

**Comment by ioquatix** [comment]
One other aspect of this that needs to be considered, is whether we also enforce this on the server side. As, in, right now you can send such a request with pseudo-headers interleaved, and it won't be rejected.

**Comment by ioquatix** [comment]
If, and it's a big if, we can do both client side and server side in this protocol gem, AND it's specced/tested, AND it has a minimal overhead, I'd be okay with it.

**Comment by ioquatix** [comment]
Wise words: "Strict with what you send, generous with what you receive".

**Comment by igrigorik** [comment]
> I still feel strongly that it is more useful to handle this outside the gem (because there might be situations where the flexibility is useful).

I don't follow. If it's against the spec, what's the flexibility? 

I don't see any reason why we can't enforce the right logic both on client and server: we can apply the right sorting on users behalf for the client, and validate received headers for the server. Given that, I think we're in agreement that we can handle this in the lib?

**Comment by ioquatix** [comment]
How do you test a remote server follows the spec if there isn’t appropriate layering of functionality that lets you send invalid requests?

**Comment by igrigorik** [comment]
Perhaps I'm missing something, but we can mock and feed anything we want into our specs, right? What stops us from testing that we issue a goaway if the received headers are in wrong order?

**Comment by ioquatix** [comment]
> Perhaps I'm missing something, but we can mock and feed anything we want into our specs, right? What stops us from testing that we issue a goaway if the received headers are in wrong order?

I'm not talking about testing `http-2`. I'm talking about writing a test in `async-http` that correctly confirms the client and server behaviour. If I can't generate malformed requests and responses, I can't test it.

There is no benefit to `async-http` in applying this patch. And so far, the implications are worse performance and less flexibility.

To be frank, the RFC behaviour regarding the order of headers doesn't affect `async-http` in any way. In fact, it's lenient in what it accepts, and strict in what it sends. Which is a reasonable stance to take. I might consider adding a warning if the server receives headers that are out of order.

More generally, there are plenty of behaviours which are not prevented by `http-2` but are specified by the RFC (for example, sending multiple header frames, and up until recently the behaviour regarding the client settings). I don't think it's reasonable for a protocol level library to implement the entire RFC especially when it comes to very high level concerns. I believe it should be allowed and accepted that users can generate malformed requests using this library, and that's okay. Sometimes that's exactly what you want to do.

**Comment by ioquatix** [comment]
Here is the correct fix.

Add .sort to the following line.

https://github.com/ostinelli/net-http2/blob/1b55c0b91d5508cb2567b24ffb46e9a5c7084772/lib/net-http2/stream.rb#L82

This problem will be solved, there will be no performance regressions, and nothing else will break.

**Comment by byronformwalt** [comment]
@ioquatix I disagree with you in principle.  It seems as though you suggesting that it is not appropriate for the http-2 gem to be compliant with the specifications because it would require that you modify tests in your code that will break once the http-2 gem becomes compliant with the spec.  I feel strongly that the correct course of action is to ensure compliance with specifications whenever possible.  Maybe I have misunderstood the purpose of this gem, but I thought it was to be a pure Ruby implementation of the HTTP/2 protocol.

@HoneyryderChuck Clearly, your solution to detecting the out-of-order header issue is more computationally efficient than to mine was.  Thanks for suggesting it.

@igrigorik How would you like to proceed (sort headers, fail by exception, or remain non-compliant)?  One of us can open a subsequent PR, referencing this one.  The most difficult part will be updating existing Rspec tests and writing new ones.

**Comment by HoneyryderChuck** [comment]
just my two cents: the library could (not should) validate the user input and at best not mangle it. Just to give a precedent: the RFC also specifies that header fields MUST be downcased, and the library is not enforcing it (it is expected that user code will take care of it). If and if it does it, it should be at least efficient in doing so. 

@byronformwalt I can come up with a fix whenever I find the time, but if you go the idea of my suggestion and you think that you have the time to implement it, let me know.

**Comment by ioquatix** [comment]
> just my two cents: the library could (not should) validate the user input and at best not mangle it. Just to give a precedent: the RFC also specifies that header fields MUST be downcased, and the library is not enforcing it (it is expected that user code will take care of it). If and if it does it, it should be at least efficient in doing so.

I strongly agree with this.

> I disagree with you in principle. It seems as though you are suggesting that it is not appropriate for the http-2 gem to be compliant with the specifications because ... strongly that the correct course of action is to ensure compliance with specifications whenever possible

I don't think you understand my position.

If you feed the valid data into this gem, it will comply with the specifications. If you feed in invalid data, there is no guarantee it will comply with the specifications. Period.

There is no way we can catch every situation like this and still be efficient; and there are lots of them.

A gem like this is not designed to baby-sit the higher level code. It should state it's invariants, and then it's up to the caller to respect them. I'd be happy with adding documentation such as "Pseudo-headers must come first otherwise the behaviour is non-conformant". It sounds like we should also add "Header keys must be lower case". As I suggested, adding a layer which sits on top which is only present during unit testing isn't a stupid approach. It can detect non-compliant use of the API but with no overhead in production.

**Comment by byronformwalt** [comment]
Just so that we all have the same info:

Section [8.1.2.1](https://http2.github.io/http2-spec/#rfc.section.8.1.2.1):

> All pseudo-header fields MUST appear in the header block before regular header fields. Any request or response that contains a pseudo-header field that appears in a header block after a regular header field MUST be treated as malformed.

Section [8.1.2.6](https://http2.github.io/http2-spec/#rfc.section.8.1.2.6):

> Malformed requests or responses that are detected MUST be treated as a stream error (Section 5.4.2) of type PROTOCOL_ERROR.

My interpretation of this information is straightforward:  If the goal of this gem is to implement the HTTP/2 protocol, both the client and server implementations at the protocol level MUST detect and treat out-of-order headers as malformed because this situation is defined as a protocol-level error (not as a user-level error) by the specification.

**Comment by ioquatix** [comment]
Thanks for pasting the spec.

I suggest adding this to the documentation of the `#headers` function in the first instance.

```
# All pseudo-header fields MUST appear in the header block before regular header fields.
```

This is already being fixed upstream here, they are aware of the issue.

https://github.com/ostinelli/net-http2/issues/21

**Comment by HoneyryderChuck** [comment]
@byronformwalt agreed. Change my "could" to "should", at least on the receiving end (although we should also validate input when sending).

**Comment by byronformwalt** [comment]
@HoneyryderChuck I may be able to work on it later in the week/weekend, if you get to it first, let me know.  I was planning to open a new PR referencing this one once @igrigorik’s approved the direction he wants to take this.

**Comment by ioquatix** [comment]
I took a look at Go's http2 implementation. They opted for the "comment" approach:

https://github.com/golang/net/blob/ed29d75add3d7c4bf7ca65aac0c6df3d1420216f/http2/frame.go#L1350-L1369

I'd be interested to see what other libraries do.

**Comment by ioquatix** [comment]
Until we find a consensus here, I've documented the requirements for conformance on the `Stream#headers` method. I feel some level of frustration here, but at the very least it's probably good to document existing behaviour/requirements.

**Comment by ioquatix** [comment]
I also feel like we should at least document the server side too... that it preserves the order of headers as provided by the client.

**Comment by ostinelli** [comment]
> Personally, I think we should handle this in http-2. This is a non-obvious requirement deep in the guts of the protocol and we should do the right thing (tm) on user's behalf.

100% agree with this. The `http-2` gem is a "pure Ruby implementation of HTTP/2 protocol", and what is this headers' ordering if not a requirement of the HTTP/2 protocol. On top of this, what @ioquatix is stating is that *every* client using this gem would have to re-implement the very same thing. These are two very strong arguments in favour of adding the proper sorting in `http-2`.

I do not understand the argument of not adding this functionality here so that `async-http` can properly be tested.

**Comment by ioquatix** [comment]
> I do not understand the argument of not adding this functionality here so that async-http can properly be tested.

If you sort the users input data, that makes it impossible to supply bad data. So you can't test the bad case.

**Comment by ostinelli** [comment]
You are not supposed to test it. `http-2` is supposed to cover that case, and if you use this gem you should trust its test suite.

**Comment by ioquatix** [comment]
> You are not supposed to test it. http-2 is supposed to cover that case, and if you use this gem you should trust its test suite.

This is a low level gem, not a high level gem. It personally think it should be possible to make, say, an http conformance checking tool using this gem.

**Comment by ioquatix** [comment]
Perhaps it would be instructive to see what other libraries do.

@ostinelli Does Erlang do this in it's low level HTTP/2 library?

**Comment by ostinelli** [comment]
> @ostinelli Does Erlang do this in it's low level HTTP/2 library?

There isn't one, only user-generated ones (as for Ruby).

**Comment by kenichi** [comment]
FWIW i think [h2](https://github.com/kenichi/h2/blob/master/lib/h2/client.rb#L106-L114) is ok in this regard, and perhaps looking at [cowboy](https://github.com/ninenines/cowboy/blob/master/src/cowboy_http2.erl) may help? i don't read erl so... /shrug

**Comment by ioquatix** [comment]
Here are the take away points so far:

- The design of existing libraries can help inform our decision.
- We don't want to affect performance or manipulate user data significantly.
- Some higher level libraries appear working correctly (`h2`, `httpx`, `async-http`) and some might need more attention (`net-http2`).

What has been done thus far:

- Document the requirement of the current code base that pseudo-headers need to come first.

Three main options to consider:

- PR: Validate order of headers when sending them and throw exception.
- PR: Sort/partition headers before sending them (could be opt-in or opt-out).
- Do nothing.

What needs to happen next:

- Implement options above, discussion. If no obvious consensus is reached, we can have private ballot, one vote per library author/concerned person.

**Comment by ioquatix** [comment]
@kenichi Thanks for your input, it's valuable to know everyone who has a stake in this issue.

**Comment by byronformwalt** [comment]
@igrigorik  Here are some current statistics on the user base of your gem that I would recommend factoring into your decision-making process for way ahead:

[async-http](https://rubygems.org/gems/async-http): 

- 9,700 total downloads
- 195 downloads for current version

[net-http2](https://rubygems.org/gems/net-http2):

- 562,162 total downloads
- 21,168 downloads for current version

In my experience, it's usually, but not always better to give more influence to the majority of users.

**Comment by ioquatix** [comment]
@byronformwalt This isn't a fight, my friend, this is about finding the best decision for everyone.

**Comment by ioquatix** [comment]
I am going to give everyone some time to think about this. I believe it's clear from https://github.com/igrigorik/http-2/pull/132#issuecomment-402314973 that we have a way forward. Please make your respective PRs and we can continue the discussion once we have something actionable.

**Comment by igrigorik** [comment]
Hey folks. First off, thanks everyone for your thoughts and input.

Stepping back, the core question we should answer and converge on first is: _what is the expected role and behavior of `http-2` against normative requirements of HTTP/2 protocol?_

My original intent, when I started working on this library, was to provide a library that enforces **all** normative `MUST` and `MUST NOT`'s (and [their equivalents](https://tools.ietf.org/html/rfc2119)) specified by the HTTP/2 RFC. There may be cases where the library might also pick a side for a `SHOULD`,  `SHOULD NOT`, and `MAY`, but generally those decisions can and ought to be left to upstream consumers.

As a result, the existing code goes to great lengths to validate client and server provided data, keeps tabs on flow control invariants, and provides built-in stream and connection error handling. Personally, I believe that we _should_ continue enforcing this and handle any identified gaps in http-2. 

As a corollary to the above, it does mean that http-2 makes it impossible to generate non-compliant h2 behaviors — i.e. it _does_ validate and act on received input both from client and server.

---

Assuming you agree with the above, in the context of header sorting: it is a MUST (https://github.com/igrigorik/http-2/pull/132#issuecomment-402052642) in the h2 RFC, which means that `http-2` needs to validate and define a behavior if the validation fails.

As a brief aside, our compressor logic already downcases header names: https://github.com/igrigorik/http-2/blob/08d1a3dcade1ae39f4598e0d4e1622fde13bb014/lib/http/2/compressor.rb#L198-L206

An alternative behavior here would have been to validate the casing and if a non-downcased name was found, then raise an error to the client and abort the request. However, we made the call to apply the transform because (a) it is safe, (b) it abstracts protocol details that most upstream clients should not concern themselves with. 

Extending the same logic, I would propose that we apply the same logic to header sorting. There is only one h2 compliant ordering of headers, sorting is similarly 'safe' and does not affect semantics of the request, and we can enforce this invariant in http-2 — there's no reason to ask every upstream library to implement same logic.

Thoughts?

**Comment by ioquatix** [comment]
@igrigorik I locked this discussion. So only users who have project-level permission will be able to reply.

Here is a PR I opened for further discussion: https://github.com/igrigorik/http-2/pull/134 and everyone is invited to contribute on that specific point.

Rather than make this thread longer, I will reply on that PR.

**Comment by HoneyryderChuck** [comment]
@igrigorik the comment about the downcasing was mine, my bad, I might have mislead someone.

I don't think that people disagree about what to do on the server side. It's the client side that has been the subject of discussion.

My 2 cents: I think that validation would be better than sorting headers. Both solutions would be ok from the spec compliance point of view, but sorting does a bit more than putting pseudo-headers before other headers. Besides the performance argument, which should not be ignored and has been discussed previously, there's also the fact that always ordering headers (also the non-pseudo) can lead malicious headers to identify the parser lib and abuse it if there are known flaws. I know that curl has a bullet point in their TODO list to "randomize" headers, to make malicious actors's lives more difficult. But on the other hand, I'll be fine with whatever solution, as compliance with spec should be prioritized. 

@ioquatix I'm not in favour of the strict flag. It has no tangible meaning, also regarding the spec, so it just makes using the lib more difficult. 

Also, I'm not in favour of locking the conversation. This all started because @byronformwalt made a PR suggested by @igrigorik , he actually heard our complaints enough to close his own PR and was willing to make a new one based on the suggestions, didn't offend anyone, and AFAIK he's the only one commenting who isn't a collaborator. Hopefully he'll still be willing to contribute.

**Comment by igrigorik** [comment]
> So only users who have project-level permission will be able to reply.

Oh, missed this. FWIW, I'd prefer not to fragment discussion over multiple threads — we already have all the participants here and previous discussion context. I'll reopen this.

> @HoneyryderChuck: I don't think that people disagree about what to do on the server side. It's the client side that has been the subject of discussion.

Great. Then first resolution here is... 

**AI 1: Implement validation (order and casing) of received headers. Failure must trigger a stream error (PROTOCOL_ERROR).**

> @HoneyryderChuck: My 2 cents: I think that validation would be better than sorting headers. Both solutions would be ok from the spec compliance point of view, but sorting does a bit more than putting pseudo-headers before other headers. Besides the performance argument, which should not be ignored and has been discussed previously, there's also the fact that always ordering headers (also the non-pseudo) can lead malicious headers to identify the parser lib and abuse it if there are known flaws. I know that curl has a bullet point in their TODO list to "randomize" headers, to make malicious actors's lives more difficult. But on the other hand, I'll be fine with whatever solution, as compliance with spec should be prioritized.

@byronformwalt suggested partitioning (https://github.com/igrigorik/http-2/pull/132#issuecomment-401672404) which addresses above concerns: it's cheaper than sorting and it allows upstream to apply whatever ordering they like. I think this is a good solution and I'd like to propose that as the way forward.

**AI 2: Header compressor should emit pseudo-headers ahead of the rest.**

---

Responding to @ioquatix's points in https://github.com/igrigorik/http-2/pull/134...

> That's why I think strict: is a good compromise. For the majority of users, you can enable it and it will work in a way which is less efficient but guarantees conformance.

I hear your point that for "compliant upstream libraries" the extra downcase / partitioning may not be necessary, but I disagree that most users will generate compliant requests by default. What we're discussing here is down-in-the-weeds details of the protocol, you average user is none the wiser to this and should not be subjected to digging through three layers of stream errors to figure out why their requests are erroring out. Conversely, I see no reason why every library should be enforcing own downcasing and partitioning — defer that to http2.

That said, _IF_ there is strong interest from other library implementers on this thread, I think we can define a contract that avoids duplicating work. Specifically, we can define a formal interface (`HTTP2::Header::List`, or some such) in http-2. If the passed in list is of that class, we can pass it right through, and if not then we'll mint it by applying appropriate transforms first.

> You are not totally correct on this point. If the sort is unstable (which is why if anything should use #partition), it might change the relative order of headers with the same key which CAN change the semantics of a request, which is why hash tables are generally a terrible idea for headers.

Per HTTP spec, headers can be re-ordered and combined at will. It is up to each header specification to define how to process merges and ordering. 

---

- Is everyone on board with AI 1 and 2? 
- Is there interest in leveraging HTTP2::Header::List upstream of http-2?

**Comment by HoneyryderChuck** [comment]
Partition might be cheaper than sort, but it's still not as cheap as ideal. However, for all the reasons stated above, I agree with the proposed solutions.

Regarding the Header::List: I don't think that it brings that much value. Any upstream library could easily create a headers "quack" which forgoes these expensive operations. But what do you think about passing these flags to #headers? Like, some :already_ordered or :ordered that informs that the input is indeed already ordered.

**Comment by igrigorik** [comment]
> But what do you think about passing these flags to #headers? Like, some :already_ordered or :ordered that informs that the input is indeed already ordered.

If I'm user of your library, it would also require that you plumb through similar flags on your end, and flags feels like a strictly worse (weaker) contract than agreeing on an interface?

> Partition might be cheaper than sort, but it's still not as cheap as ideal. However, for all the reasons stated above, I agree with the proposed solutions.

Great!

@ostinelli @byronformwalt @ioquatix are you on board with proposal outlined above?

**Comment by byronformwalt** [comment]
It’s a good compromise that prevents unnecessary overhead for users that have taken measures to presort headers IAW the protocol spec.

**Comment by ostinelli** [comment]
> Is everyone on board with AI 1 and 2?

This is consistent with what I expressed, so fine by me. Will re-release another version of `net-http2` when you do so, since I meanwhile had to patch it with a simplified sorting of headers.

> Is there interest in leveraging HTTP2::Header::List upstream of http-2?

It's a good suggestion, but not needed by me.

**Comment by igrigorik** [comment]
I think we're zeroing in on a solution here. As a recap..

> AI 1: Implement validation (order and casing) of received headers. Failure must trigger a stream error (PROTOCOL_ERROR).
> AI 2: Header compressor should emit pseudo-headers ahead of the rest (by partitioning supplied list).

I'm not hearing strong demand for an explicit interface for headers, so I guess we can defer that. Please chime in if anyone disagrees with any of the above.

Last but not least, landing the patches.. @byronformwalt would you be up for (re)drafting a PR to cover the above behaviors? :)

**Comment by byronformwalt** [comment]
@igrigorik I want to ensure I understand what is to be done.  Do you want AI 1 to be performed on the Decompressor class and while AI 2 is performed on the Compressor class?

**Comment by igrigorik** [comment]
AI 1: yeah, I think this could be done in decompressor, or alternatively in [decode_headers](https://github.com/igrigorik/http-2/blob/190830b62c2f42698ae6a930022ce59d3bad7378/lib/http/2/connection.rb#L604). Need to think about the plumbing, since we need to return the correct error.

AI 2: yes, I think we should handle that in Compressor, same as downcasing.

**Pull request closed without merging**